independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Types.java

```diff
@@ -277,20 +279,37 @@ static Type getProviderType(ClassInfo classInfo) {
         return types;
     }
 
-    static <T extends Type> Map<TypeVariable, Type> buildResolvedMap(List<T> resolvedTypeVariables,
+    static <T extends Type> Map<TypeVariable, Type> buildResolvedMap(List<T> resolvedArguments,
             List<TypeVariable> typeVariables,
-            Map<TypeVariable, Type> resolvedTypeParameters) {
+            Map<TypeVariable, Type> resolvedTypeParameters, IndexView index) {
         Map<TypeVariable, Type> resolvedMap = new HashMap<>();
-        for (int i = 0; i < resolvedTypeVariables.size(); i++) {
-            Type resolvedTypeVariable = resolvedTypeVariables.get(i);
-            Type resolvedTypeParam = Kind.TYPE_VARIABLE.equals(resolvedTypeVariable.kind())
-                    ? resolvedTypeParameters.getOrDefault(resolvedTypeVariable, resolvedTypeVariable)
-                    : resolvedTypeVariable;
-            resolvedMap.put(typeVariables.get(i), resolvedTypeParam);
+        for (int i = 0; i < resolvedArguments.size(); i++) {
+            resolvedMap.put(typeVariables.get(i), resolveTypeParam(resolvedArguments.get(i), resolvedTypeParameters, index));
         }
         return resolvedMap;
     }
 
+    static Type resolveTypeParam(Type typeParam, Map<TypeVariable, Type> resolvedTypeParameters, IndexView index) {
+        if (typeParam.kind() == Kind.TYPE_VARIABLE) {
+            return resolvedTypeParameters.getOrDefault(typeParam, typeParam);
+        } else if (typeParam.kind() == Kind.PARAMETERIZED_TYPE) {
+            ParameterizedType parameterizedType = typeParam.asParameterizedType();
+            ClassInfo classInfo = index.getClassByName(parameterizedType.name());
+            if (classInfo != null) {
```

Comment:
If the class isn't on index, shouldn't we throw an exception or at least log something?
An injection scenario relying on this structure and not having the class indexed would give user an hard-to-read unsatisfied dependency then.

Reply:
That's a good point. However, it should be probably addressed in a separate issue because it's a more general problem. Currently, our `io.quarkus.arc.processor.BeanArchives.IndexWrapper` attempts to index classes that are not part of the index. As a result, it's very unlikely that `null` is returned.

Reply:
@mkouba https://github.com/quarkusio/quarkus/issues/4080 feel free to drop a comment.